lib/dataSource-url.js

```diff
@@ -202,6 +205,24 @@ kettle.dataSource.URL.errorCallback = function (promise, whileMsg) {
     };
 };
 
+kettle.dataSource.prepareLoggingOptions = function (requestOptions, toCensor) {
+    var sensitiveValues = fluid.values(fluid.transform(toCensor, function (value, key) {
+        return value ? requestOptions[key] : null;
+    }));
+
+    var censorURL = function (url) {
+        sensitiveValues.forEach(function (sensitiveValue) {
+            if (sensitiveValue) {
+                url = url.replace(sensitiveValue, "[CENSORED]");
```

Comment:
This straightforward string replacement will fail whenever there is percent-encoding in the URL. For example, if we change the test URL in file `tests/DataSourceSimpleTests.js` from `"https://secret-user:secret-password@thing.available:997/path"` to `"https://secret-user:secret-%25-password@thing.available:997/path"` (that is, a password with value "secret-%-password"), the test will fail.

In this case, the full URL is logged as `"https://secret-user:secret-%25-password@thing.available:997/path"`.

Inside the `censorURL` function, `sensitiveValue` is decoded and has the value `"secret-user:secret-%-password"`. Which will not be found in the URL for replacement.

Reply:
Is it feasible to never log the URL?

I fear that this approach will be difficult to get right as there are multiple possible notations that result in the same decoded value. For example:

- AB
- %41%42

Now this example is unlikely to occur (in practice, we will always get AB) but the possibility exists perhaps for some other characters (some punctuation maybe or perhaps non-Latin characters).

Reply:
Thanks for spotting this risk, @simonbates - I decided the readability advantages of logging the URL warranted keeping it, and changed the workflow to regenerate the URL by re-encoding the already censored broken-down fields rather than attempting to censor it in place. Ready for another look - cheers